metricflow/inference/context/data_warehouse.py

```diff
@@ -0,0 +1,116 @@
+    max_value: Optional[T]
+
+    def __post_init__(self, dtype: Optional[str]):  # noqa: D
+        object.__setattr__(self, "type", SqlColumnType.from_pandas_dtype(dtype))
+
+    @property
+    def is_empty(self) -> bool:
+        """Whether the column has any rows"""
+        return self.row_count == 0
+
+    @property
+    def cardinality(self) -> float:
+        """The proportion between unique values and the row count of the column."""
+        if self.is_empty:
+            raise InferenceError("Cannot determine the cardinality of empty column.")
+
+        return self.distinct_row_count / self.row_count
+
+    @property
+    def is_nullable(self) -> bool:
+        """Whether the column is nullable or not"""
+        return self.null_count != 0
+
+
+@dataclass(frozen=True)
+class TableStatistics:
+    """Holds statistical data about a table."""
+
+    column_stats: InitVar[List[ColumnStatistics]]
+
+    table: SqlTable
+    columns: Dict[SqlColumn, ColumnStatistics] = field(default_factory=lambda: {}, init=False)
+
+    def __post_init__(self, column_stats: List[ColumnStatistics]) -> None:  # noqa: D
+        for col in column_stats:
+            self.columns[col.column] = col
+
+
+@dataclass(frozen=True)
+class DataWarehouseInferenceContext(InferenceContext):
+    """The inference context for a data warehouse. Holds statistics and metadata about each column."""
+
+    table_stats: InitVar[List[TableStatistics]]
+
+    tables: Dict[SqlTable, TableStatistics] = field(default_factory=lambda: {}, init=False)
+    columns: Dict[SqlColumn, ColumnStatistics] = field(default_factory=lambda: {}, init=False)
+
+    def __post_init__(self, table_stats: List[TableStatistics]) -> None:  # noqa: D
+        for stats in table_stats:
+            self.tables[stats.table] = stats
+            for column in stats.columns.values():
+                self.columns[column.column] = column
+
+
+@dataclass(frozen=True)
+class DataWarehouseInferenceContextProvider(InferenceContextProvider[DataWarehouseInferenceContext]):
+    """Provides inference context from a data warehouse by querying data from its tables.
+
+    client: the underlying SQL engine client that will be used for querying table data.
+    tables: an exhaustive list of all tables that should be queried.
+    max_sample_size: max number of rows to sample from each table
+    """
+
+    client: SqlClient
+    tables: List[SqlTable]
+    max_sample_size: int = 1000
+
+    def _get_table_statistics(self, table: SqlTable) -> TableStatistics:
+        """Fetch statistics about a single table by querying the warehouse."""
+
+        query = f"SELECT * FROM {table.sql} LIMIT {self.max_sample_size}"
+        df = self.client.query(query)
+        column_stats = [
+            ColumnStatistics(
+                column=SqlColumn(table=table, name=col_name),
+                dtype=str(series.dtype),
+                row_count=len(series),
+                distinct_row_count=series.nunique(dropna=False),
```

Comment:
This is going to produce the wrong value. You can't do a distinct on a non-hash sample (and *especially* not on a limit sample, since those rows might be ordered on insert) and make any claims at all. Imagine a table that has 10,000 distinct values in this column 10 of which represent 99.99% of rows, or one with 10,000 evenly distributed values that was insert ordered on that column, or one with a time dependency on a much-more-common time series insertion order. 

In the more common cases, randomly distributed moderate cardinality columns are going to produce very high variance (i.e., completely unreliable) output with this strategy, even when normalized against input row count.

I know this is for a by-design-constraint imperfect inference step, but this seems like it'll produce a lot more noise than we want. I recommend leaving this out altogether for the time being, or else doing it via a separate column sampling pass only for cases where it seems warranted (or where the user requests us to do the relevant set of inferences, or something).

Reply:
**About PANDAS usage:**
I was actually going for the native SQL approach for a couple of reasons:
1. We're not constrained by the client's memory and could theoretically submit all the queries at once to be run natively at the warehouse. I think this could be a neat performance improvement for the future.
2. The data warehouse is built exactly for querying, and is probably much more reliable and fast than we can probably ever get with a naive client-side approach.

I was initially planning on running something like this:
```sql
with source as (
    select * from DEMO_DB.CUSTOMERS limit 10000
    -- possibly some other fancy sampling strategy here
)

select 
    count(*) as count, 
    
    count(distinct customer_id) as customer_id_count,
    min(customer_id) as customer_id_min,
    max(customer_id) as customer_id_max,
    sum(case when customer_id is null then 1 else 0 end) as customer_id_countnull,
    
    count(distinct ds) as ds_count,
    min(ds) as ds_min,
    max(ds) as ds_max,
    sum(case when ds is null then 1 else 0 end) as ds_countnull
from
    source
```

However, I thought warehouses could get kinda finicky with ANSI SQL and for querying stuff like `MIN` for a nested array field, JSON or other weird field. I decided for running it all client-side just to get a thing going for V0. Do you think that's a reasonable decision or should we just switch to native SQL from the start?

**About sampling**
I was hoping with a sufficiently large N, the sample distribution of nulls, different values etc would mimic the real distribution in the data. I guess this is not the case for time series data that's inserted in order. I'm also aware this could be very problematic for MIN/MAX values, since random samples are very likely to get those values right.

I do think we may still want to employ some sampling strategy, though. My main concern with this is running something like `COUNT(DISTINCT user_id)` over a huge table (terabytes or petabytes) will for sure get _very_ expensive.

Honestly, I don't know what is a good solution to this for now.

Reply:
I'm resolving this since we've moved to a warehouse-native approach, which gets rid of client-side Pandas and uses Snowflake's native `SAMPLE` capabilities.

I'll definitely watch out for noise in the data once we're implementing all the inference rules and policies. I guess the only way to estimate a tolerable amount of noise is to actually run inference over tables and see how this performs.